src/beast/beast/threads/Stoppable.h

```diff
@@ -314,22 +324,56 @@ class RootStoppable : public Stoppable
         Thread safety:
             Safe to call from any thread not associated with a Stoppable.
     */
-    void stop (Journal journal = Journal());
+    void stop (Journal j);
+
+    /** Sleep or wake up on stop.
+
+        @return `true` if we are stopping
+    */
+    template <class Rep, class Period>
+    bool
+    alertable_sleep_for(
+        std::chrono::duration<Rep, Period> const& d);
+
 private:
-    /** Notify a root stoppable and children to stop, without waiting.
+    /*  Notify a root stoppable and children to stop, without waiting.
         Has no effect if the stoppable was already notified.
 
         Thread safety:
             Safe to call from any thread at any time.
     */
-    void stopAsync ();
+    void stopAsync(Journal j);
 
     std::atomic<bool> m_prepared;
     std::atomic<bool> m_calledStop;
     std::atomic<bool> m_calledStopAsync;
+    std::mutex m_;
+    std::condition_variable c_;
 };
 /** @} */
 
+//------------------------------------------------------------------------------
+
+template <class Rep, class Period>
+bool
+RootStoppable::alertable_sleep_for(
+    std::chrono::duration<Rep, Period> const& d)
+{
+    std::unique_lock<std::mutex> lock(m_);
+    if (m_calledStop)
+        return true;
+    return c_.wait_for(lock, d) ==
+        std::cv_status::no_timeout;
```

Comment:
Without a predicate, `wait_for` can have spurious wakeups. Given this is a sleep function, and looking at the callers, this looks OK to me. However, a spurious wakeup would make the return value incorrect. I'd suggest two changes:

1) Change the return type to `void`
2) Add a comment `// spurious wakeups OK` by the `wait_for` function.